File: app/src/main/java/lt/zilinskas/marius/testio/fragments/ServersFragment.java
package lt.zilinskas.marius.testio.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import lombok.Getter;
import lombok.Setter;
import lt.zilinskas.marius.testio.R;
import lt.zilinskas.marius.testio.adapters.ServersAdapter;
import lt.zilinskas.marius.testio.api.entities.Server;
import lt.zilinskas.marius.testio.database.DatabaseHelper;

public class ServersFragment extends BaseFragment {

    private CustomViewHolder customViewHolder;
    private CompositeDisposable compositeDisposable;
    private LoadedData loadedData;
    private ServersAdapter serversAdapter;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        compositeDisposable = new CompositeDisposable();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_servers, container, false);
        customViewHolder = new CustomViewHolder(view);

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        loadData();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (compositeDisposable != null && !compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
    }

    private void loadData() {
        Disposable disposable = Observable
                .just(testioApplication.getDatabaseHelper())
                .map(this::loadDataFromDb)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::dataLoaded);
        compositeDisposable.add(disposable);
    }

    private LoadedData loadDataFromDb(DatabaseHelper databaseHelper) {
        loadedData = new LoadedData();
        loadedData.setServers(databaseHelper.getExpensesDAO().getAllServers());
        return loadedData;
    }

    private void dataLoaded(LoadedData loadedData) {
        this.loadedData = loadedData;
        populateAdapter();
    }

    private void populateAdapter() {
        if (serversAdapter != null && customViewHolder.serversList.getAdapter() != null) {
            serversAdapter.setServers(loadedData.servers);
            serversAdapter.notifyDataSetChanged();
            return;
        }

        initAdapter();
    }

    private void initAdapter() {
        if (getContext() == null) {
            return;
        }
        serversAdapter = new ServersAdapter(loadedData.servers);

        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getContext());
        customViewHolder.serversList.setHasFixedSize(true);
        customViewHolder.serversList.addItemDecoration(new DividerItemDecoration(getContext(),
                DividerItemDecoration.VERTICAL));
        customViewHolder.serversList.setLayoutManager(layoutManager);
        customViewHolder.serversList.setItemAnimator(new DefaultItemAnimator());
        customViewHolder.serversList.setAdapter(serversAdapter);
    }


    @Getter
    @Setter
    private class LoadedData {
        private List<Server> servers;
    }

    @Getter
    private class CustomViewHolder {
        private View parent;
        private RecyclerView serversList;

        CustomViewHolder(@NonNull View parent) {
            this.parent = parent;
            initViews();
        }

        private void initViews() {
            serversList = parent.findViewById(R.id.serversList);
        }
    }
}

File: app/src/main/java/com/giedrius/androidparty/task/view/ServersFragment.kt
package com.giedrius.androidparty.task.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.giedrius.androidparty.R
import com.giedrius.androidparty.task.adapter.ServerAdapter
import com.giedrius.androidparty.task.navigation.NavigationListener
import com.giedrius.androidparty.task.viewmodel.ServersListViewModel
import dagger.android.support.AndroidSupportInjection
import dagger.android.support.DaggerFragment
import kotlinx.android.synthetic.main.fragment_servers.view.*
import javax.inject.Inject

class ServersFragment(val navListener: NavigationListener): DaggerFragment() {

    @Inject lateinit var serversViewModel: ServersListViewModel
    lateinit var serversAdapter: ServerAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AndroidSupportInjection.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = LayoutInflater.from(context).inflate(R.layout.fragment_servers, container, false)
        view.server_list_recycler.layoutManager = LinearLayoutManager(activity)
        serversAdapter = ServerAdapter(listOf())
        view.server_list_recycler.adapter = serversAdapter
        serversViewModel.getServers()
        view.server_list_recycler.addItemDecoration(DividerItemDecoration(activity, DividerItemDecoration.VERTICAL))
        view.logout_button.setOnClickListener { navListener.onLogoutPerformed() }
        createServersObserver()
        return view
    }

    private fun createServersObserver() {
        serversViewModel.servers.observe(this, Observer {
            serversAdapter.setData(it.sortedBy { it.distance })
            serversAdapter.notifyDataSetChanged()
        })
    }

    companion object {
        @JvmStatic fun newInstance(navlistener: NavigationListener) = ServersFragment(navlistener).apply{}
    }
}
